src/allmydata/storage_client.py

```diff
@@ -1008,6 +1007,54 @@ def try_to_connect(self):
         self._reconnector.reset()
 
 
+def _pick_a_http_server(
```

Comment:
As discussed on IRC, I think the implementation of `race` in https://github.com/twisted/twisted/pull/11818 does most of what `_pick_a_http_server` does and has a couple fewer bugs relative to this version.  Let's adopt the version from that PR (and then `_pick_a_http_server` becomes a thin wrapper around `race`).